transaction/LockManager.cpp

```diff
@@ -0,0 +1,235 @@
+      lock_table_->unlatchExclusive();
+      return false;
+    }
+    stack.pop();
+  }
+  lock_table_->unlatchExclusive();
+  return true;
+}
+
+bool LockManager::releaseAllLocks(const transaction_id tid) {
+  std::vector<ResourceId> related_rids
+      = transaction_table_->getResourceIdList(tid);
+  TransactionTableResult transaction_deleted
+      = transaction_table_->deleteTransaction(tid);
+  if (transaction_deleted
+          == TransactionTableResult::kTransactionDeleteError) {
+    FATAL_ERROR("In LockManager.releaseAllLocks"
+                " transaction could not be deleted!");
+  }
+
+  lock_table_->latchExclusive();
+
+  for (std::vector<ResourceId>::const_iterator it = related_rids.begin();
+       it != related_rids.end();
+       ++it) {
+    LockTableResult lock_deleted = lock_table_->deleteLock(tid, *it);
+
+    std::cout << "Transaction " + std::to_string(tid)
+      + " released lock:" + it->toString() + "\n";
+
+    if (lock_deleted == LockTableResult::kDelError) {
+      FATAL_ERROR("In LockManager.releaseAllLock "
+                  "lock could not be deleted from LockTable");
+    }
+  }
+
+  lock_table_->unlatchExclusive();
+  return true;
+}
+
+// If not blocked return true
+// if block return false
+bool LockManager::acquireLockInternal(const transaction_id tid,
+                                      const ResourceId &rid,
+                                      const AccessMode access_mode) {
+  LockTableResult l_result = lock_table_->putLock(tid, rid, access_mode);
+  if (l_result == LockTableResult::kPutError) {
+    LOG(FATAL) << "Unexpected result in LockManager.acquireLockInternal";
+  }
+
+  if (l_result == LockTableResult::kAlreadyInOwned) {
+    return true;
+  } else if (l_result == LockTableResult::kPlacedInOwned) {
+    TransactionTableResult t_result
+        = transaction_table_->putOwnEntry(tid, rid, access_mode);
+    if (t_result != TransactionTableResult::kPlacedInOwned) {
+      LOG(FATAL) << "Unexpected result in LockManager.acquireLockInternal: "
+                    "Mismatch of table results (No.1).";
+    }
+    return true;
+  } else if (l_result == LockTableResult::kAlreadyInPending) {
+    return false;
+  } else if (l_result == LockTableResult::kPlacedInPending) {
+    TransactionTableResult t_result =
+      transaction_table_->putPendingEntry(tid, rid, access_mode);
+    if (t_result != TransactionTableResult::kPlacedInPending) {
+      LOG(FATAL) << "Unexpected result in LockManager.acquireLockInternal: "
+                    "Mismatch of table results (No.2).";
+    }
+    return false;
+  }
+  return false;
+}
+
+void LockManager::killVictims() {
+  // TODO(Hakan): Find a method to latch this function
+  //              (it cannot because it calls releaseLocks)
+  if (detector_status_.load() == DeadLockDetectorStatus::kDone) {
+    for (std::vector<transaction_id>::const_iterator iter
+             = victim_result_.begin();
+         iter != victim_result_.end();
+         ++iter) {
```

Comment:
How about a `for-each`: `for (const auto victim_transaction_id : victim_result_)`.